debian/sourcemapper.go

```diff
@@ -0,0 +1,170 @@
+// the Update method.
+func NewSourcesMap(release Release, opts ...option) *SourcesMap {
+	d := &SourcesMap{
+		release:    release,
+		sourcesURL: sourcesURL,
+		mu:         &sync.RWMutex{},
+		sourceMap:  make(map[string]map[string]struct{}),
+		etagMap:    make(map[string]string),
+		client:     http.DefaultClient,
+	}
+	for _, opt := range opts {
+		opt(d)
+	}
+	return d
+}
+
+// SourcesMap wraps a map that defines the relationship between a source
+// package and it's associated binaries. It offers an Update method
+// to populate and update the map. It is Release-centric.
+//
+// It should have the same lifespan as the Updater to save allocations
+// and take advantage of the entity tag that debian sends back.
+type SourcesMap struct {
+	release    Release
+	sourcesURL string
+	mu         *sync.RWMutex
+	sourceMap  map[string]map[string]struct{}
+	etagMap    map[string]string
+	client     *http.Client
+}
+
+// Get returns all the binaries associated with a source package
+// identified by a string. Empty slice is returned if the source
+// doesn't exist in the map.
+func (m *SourcesMap) Get(source string) []string {
+	if m.sourceMap[source] == nil {
+		return []string{}
+	}
+	bins := []string{}
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for bin := range m.sourceMap[source] {
+		bins = append(bins, bin)
+	}
+	return bins
+}
+
+// Update pulls the Sources.gz files for the different repos and saves
+// the resulting source to binary relationships.
+func (m *SourcesMap) Update(ctx context.Context) error {
+	return m.updateSources(ctx)
+}
+
+func (m *SourcesMap) updateSources(ctx context.Context) error {
+	if m.release == Wheezy {
+		// There are no Wheezy records we assume the source->binary relationship of Jessie.
+		m.release = Jessie
+	}
+	for _, r := range repos {
+		url := fmt.Sprintf(m.sourcesURL, m.release, r)
+		err := m.fetchSources(ctx, url)
+		if err != nil {
+			return fmt.Errorf("unable to fetch sources: %w", err)
+		}
+	}
+	return nil
+}
+
+func (m *SourcesMap) fetchSources(ctx context.Context, url string) error {
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "debian/sourcemapper.fetchSources"),
+		label.String("url", url))
+	zlog.Debug(ctx).Msg("attempting fetch of Sources file")
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	switch resp.StatusCode {
+	case http.StatusOK:
+	default:
+		return fmt.Errorf("received status code %d querying mapping url %s", resp.StatusCode, url)
+	}
+
+	newEtag := resp.Header.Get("etag")
+	if etag, ok := m.etagMap[url]; ok && etag == newEtag {
+		zlog.Debug(ctx).Str("etag", etag).Msg("already seen this etag, skipping update")
+		return nil
+	}
+	m.etagMap[url] = newEtag
+
+	var reader io.ReadCloser
+	switch resp.Header.Get("Content-Type") {
+	case "application/x-gzip":
```

Comment:
```suggestion
	case "application/gzip", "application/x-gzip":
```

Should have the IANA type and the pre-IANA type, just for future proofing.